File: backend/internal/domain/exam.go
package domain

import (
	"time"

	"gorm.io/gorm"
)

// Exam represents an exam with multiple questions
type Exam struct {
	ID          uint           `gorm:"primarykey" json:"id"`
	Title       string         `gorm:"size:255;not null" json:"title"`
	Description string         `gorm:"type:text" json:"description"`
	Duration    int            `gorm:"not null" json:"duration"` // Duration in minutes
	ContentHTML string         `gorm:"type:text" json:"content_html"`
	IsActive    bool           `gorm:"default:true" json:"is_active"`
	Questions   []Question     `gorm:"foreignKey:ExamID" json:"questions,omitempty"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName specifies the table name for Exam
func (Exam) TableName() string {
	return "exams"
}

File: frontend/src/types/exam.ts
// Domain types matching backend models

export type QuestionType = 'multiple_choice' | 'essay' | 'true_false';

export interface Exam {
  id: number;
  title: string;
  description: string;
  duration: number; // minutes
  content_html: string;
  is_active: boolean;
  questions?: Question[];
  created_at: string;
  updated_at: string;
}

export interface Question {
  id: number;
  exam_id: number;
  content_html: string;
  type: QuestionType;
  options: any; // JSONB - flexible structure
  answer: any; // JSONB - correct answer(s)
  explanation: string;
  points: number;
  order: number;
  created_at: string;
  updated_at: string;
}

export type SessionStatus = 'in_progress' | 'completed' | 'abandoned';

export interface ExamSession {
  id: number;
  user_id: string;
  exam_id: number;
  exam?: Exam;
  start_time: string;
  end_time?: string;
  status: SessionStatus;
  score?: number;
  total_points: number;
  answers?: Answer[];
  created_at: string;
  updated_at: string;
}

export interface Answer {
  id: number;
  session_id: number;
  question_id: number;
  question?: Question;
  user_answer: any; // JSONB - user's answer
  is_correct?: boolean;
  points?: number;
  created_at: string;
  updated_at: string;
}

// Request/Response types
export interface CreateSessionRequest {
  exam_id: number;
  user_id: string;
}

export interface SubmitAnswerRequest {
  question_id: number;
  user_answer: any;
}

export interface SubmitExamResponse {
  session: ExamSession;
  score: number;
  total_points: number;
  correct_answers: number;
  total_questions: number;
}
